Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm

```diff
@@ -576,12 +576,49 @@
     CheckedPtr renderer = node.renderer();
     if (!renderer)
         return false;
+    return isTransparentOrFullyClipped(*renderer);
+}
 
-    auto* enclosingLayer = renderer->enclosingLayer();
+bool WebPage::isTransparentOrFullyClipped(const RenderObject& renderer) const
+{
+    CheckedPtr enclosingLayer = renderer.enclosingLayer();
     if (enclosingLayer && enclosingLayer->isTransparentRespectingParentFrames())
         return true;
 
-    return renderer->hasNonEmptyVisibleRectRespectingParentFrames();
+    return renderer.hasEmptyVisibleRectRespectingParentFrames();
+}
+
+static RenderObject* closestCommonContainerInRenderTree(const Position& start, const Position& end)
+{
+    RefPtr startContainer = start.containerNode();
+    if (!startContainer)
+        return nullptr;
+
+    RefPtr endContainer = end.containerNode();
+    if (!endContainer)
+        return nullptr;
+
+    CheckedPtr startRenderer = startContainer->renderer();
+    if (!startRenderer)
+        return nullptr;
+
+    CheckedPtr endRenderer = endContainer->renderer();
+    if (!endRenderer)
+        return nullptr;
+
+    if (startRenderer == endRenderer)
+        return startRenderer.get();
+
+    UncheckedKeyHashSet<CheckedPtr<RenderElement>> ancestorsOfStart;
+    for (CheckedPtr container = startRenderer->container(); container; container = container->container())
+        ancestorsOfStart.add(container);
+
+    for (CheckedPtr container = endRenderer->container(); container; container = container->container()) {
+        if (ancestorsOfStart.contains(container))
```

Comment:
Do we need to consider that `contains()` will not respect the insertion sequence of containers into `ancestorsOfStart`?

Reply:
Hm.. I think it should be fine, since `ancestorsOfStart` is unordered anyways. (This is just a way to scan for the deepest renderer containing the end that _also_ contains the start, in `O(d)` time)